### ISSUE TITLE
picoquic_add_to_stream  method

### ISSUE THREAD

**Issue by ArubaGit** [private-octopus/picoquic#1774]
Hi
I am facing an issue where I am sending string (header) of say around 70-80 bytes using this method and after this I am sending a video file using usual picoquic_provide_stream_data_buffer. This is being done for several iterations. But what is happening is that the receiver is not receiving the entire string (header) in few runs. Is there any limitation of picoquic_add_to_stream method ?

**Comment by huitema** [comment]
This is a widely tested scenario, and it ought to work. But there may be some kind of race condition going on, especially if you are in a multi-threaded environment. Please remember that the picoquic API like "add to stream" are NOT thread safe. Could it be that the video is added to the stream because picoquic_provide_stream_data_buffer is called before picoquic_add_to_stream is called? Is the stream already marked "active" when you are calling picoquic_add_to_stream?

**Comment by ArubaGit** [comment]
No picoquic_add_to_stream is called before picoquic_provide_stream_data_buffer. 
And -- Is the stream already marked "active" when you are calling picoquic_add_to_stream? -- No. It is being marked active after add_to_stream method.
/picoquic_add_to_stream(cnx, stream_id, (const uint8_t*)&filename_len, sizeof(filename_len), 0);
                        picoquic_add_to_stream(cnx, stream_id, (const uint8_t*)header, header_len, 0);
                        cout<<"sending header "<<header<<endl;

                        // Copy the filename into stream_ctx->file_name
                        strncpy((char*)stream_ctx->file_name, filename, sizeof(stream_ctx->file_name) - 1);
                        stream_ctx->file_name[sizeof(stream_ctx->file_name) - 1] = '\0';
                            
                        int stream_ret = sample_server_open_stream(server_ctx, stream_ctx);
                        if (stream_ret == 0) {
                            // If data needs to be sent, set the context as active
                            printf("marking streamID: %ld active\n", stream_id);
                            ret = picoquic_mark_active_stream(cnx, stream_id, 1, stream_ctx);
                        } else {
                            // If the file could not be read, reset the stream
                            printf("deleting\n");
                            sample_server_delete_stream_context(server_ctx, stream_ctx);
                            (void) picoquic_reset_stream(cnx, stream_id, stream_ret);
                        }
                        
  Can you confirm whether this the correct way to use both the methods?

**Comment by huitema** [comment]
I cannot confirm anything without seeing your code. I can confirm that if you call `picoquic_add_to_stream` with a stream ID, then call `picoquic_mark_active_stream` with the same stream ID, the callback to `picoquic_callback_prepare_to_send` will push data only after what `picoquic_add_to_stream` added. Unless you are calling `picoquic_add_to_stream` from inside the `picoquic_callback_prepare_to_send` callback, or with a different stream id, or maybe in a different thread.

**Comment by ArubaGit** [comment]
Hey Thanks
I sorted that out. I had one more question, Is there any way to clear the data buffered in the stream without actually resetting it.

**Comment by huitema** [comment]
I don't think so. Once the data is written, it goes out on the network, packets are queued for repeat, etc. If you want to clear all that, resetting is logical.
